### PR TITLE
OHRM5X-2480: Develop language package validation

### PR DESCRIPTION
**Pull request by TharakaAthukorala**

### CODE REVIEW

File: src/plugins/orangehrmAdminPlugin/Api/I18NTranslationImportAPI.php [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+     *                  @OA\Property(property="xliffStringValidation", description="The language strings that failed to import", type="integer"),
+     *                  @OA\Property(property="xliffFileValidation", description="The language file that failed to import")
+     *              )
+     *          )
+     *      )
+     * )
+     *
+     * @inheritDoc
+     * @throws InvalidParamException
+     * @throws \Exception
```
Comment (devishke-orange) [inline src/plugins/orangehrmAdminPlugin/Api/I18NTranslationImportAPI.php:90]: remove exceptions

```suggestion
     * @inheritDoc
```

File: src/plugins/orangehrmAdminPlugin/Service/LocalizationService.php [hunk 2]
```diff
@@ -230,23 +233,26 @@ private function getXliffXmlSources(array $i18nGroups, I18NLanguage $language):
         $root->setAttribute('srcLang', 'en_US');
         $root->setAttribute('trgLang', $language->getCode());
         $root->setAttribute('xmlns', 'urn:oasis:names:tc:xliff:document:2.0');
-        $root->setAttribute('date', @date('Y-m-d\TH:i:s\Z'));
+//        $root->setAttribute('date', @date('Y-m-d\TH:i:s\Z'));
```
Comment (devishke-orange) [inline src/plugins/orangehrmAdminPlugin/Service/LocalizationService.php:236]: date attribute is not allowed right? we can remove this line

https://docs.oasis-open.org/xliff/xliff-core/v2.0/xliff-core-v2.0.html

File: src/plugins/orangehrmAdminPlugin/Service/LocalizationService.php [hunk 3]
```diff
@@ -270,4 +276,122 @@ private function getXliffXmlSources(array $i18nGroups, I18NLanguage $language):
...
+                'line' => $xmlError['line'],
+                'column' => $xmlError['column'],
+                'message' => $xmlError['message'],
+            ];
+        }
+
+        libxml_clear_errors();
+        libxml_use_internal_errors($internal);
+
+        return ['file' => $file, 'isValid' => 0 === \count($errors), 'messages' => $errors];
```
Comment (devishke-orange) [inline src/plugins/orangehrmAdminPlugin/Service/LocalizationService.php:326]: just noting down that code taken from `XliffLintCommand` in `symfony/translations`
https://github.com/symfony/translation/blob/5.4/Command/XliffLintCommand.php#L123-L156

We'll have to discuss this a bit. Not sure if we need the filename checking part. 

Better to create a utility class for `XlfifUtils` for now. Reference: https://github.com/orangehrm/orangehrm/blob/main/src/plugins/orangehrmCorePlugin/Utility/Sanitizer.php

File: src/plugins/orangehrmAdminPlugin/Service/LocalizationService.php [hunk 4]
```diff
@@ -270,4 +276,122 @@ private function getXliffXmlSources(array $i18nGroups, I18NLanguage $language):
...
+
+        // Initialize an array to store validation errors
+        $errors = [];
+        $locale = "en_US";
+
+        // Get all the <unit> elements from the XLIFF document
+        $units = $xliffDocument->getElementsByTagName('unit');
+
+        // Define validation constraints
+        $maxLength = 255;
```
Comment (devishke-orange) [inline src/plugins/orangehrmAdminPlugin/Service/LocalizationService.php:354]: how did you decide this value?

**Review by devishke-orange — CHANGES_REQUESTED** [review]

File: src/client/src/orangehrmAdminPlugin/pages/languageImport/LanguageImport.vue [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+            </oxd-text>
+          </li>
+          <li>
+            <oxd-text class="orangehrm-information-card-text">
+              Do not change the template.
+            </oxd-text>
+          </li>
+          <li>
+            <oxd-text class="orangehrm-information-card-text">
+              Sample XLIFF :
```
[the remark this answers is not in the document]
Comment (TharakaAthukorala) [reply src/client/src/orangehrmAdminPlugin/pages/languageImport/LanguageImport.vue:48]: Added

File: src/plugins/orangehrmAdminPlugin/Api/I18NTranslationImportAPI.php [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+use OrangeHRM\Core\Api\V2\Validator\Rules;
+
+class I18NTranslationImportAPI extends Endpoint implements CollectionEndpoint
+{
+    use LocalizationServiceTrait;
+
+    public const PARAMETER_LANGUAGE_ID = 'languageId';
+    public const PARAMETER_ATTACHMENT = 'attachment';
+
+    public const PARAM_RULE_IMPORT_FILE_FORMAT = ['text/xml', 'application/xml', 'application/xliff+xml'];
```
[the remark this answers is not in the document]
Comment (TharakaAthukorala) [reply src/plugins/orangehrmAdminPlugin/Api/I18NTranslationImportAPI.php:42]: - The MIME type application/xliff+xml is the standard MIME type for .xlf files.
-  For .xlf files, it's theoretically possible for other MIME types to be associated with this extension.

File: src/plugins/orangehrmAdminPlugin/Api/I18NTranslationImportAPI.php [hunk 7]
```diff
@@ -0,0 +1,158 @@
+<?php
```
[the remark this answers is not in the document]
Comment (TharakaAthukorala) [reply src/plugins/orangehrmAdminPlugin/Api/I18NTranslationImportAPI.php:1]: added

File: src/plugins/orangehrmAdminPlugin/Api/I18NTranslationImportAPI.php (hunk 1 above)
Comment (TharakaAthukorala) [reply src/plugins/orangehrmAdminPlugin/Api/I18NTranslationImportAPI.php:90]: done

File: src/plugins/orangehrmAdminPlugin/Service/LocalizationService.php (hunk 2 above)
Comment (TharakaAthukorala) [reply src/plugins/orangehrmAdminPlugin/Service/LocalizationService.php:236]: removed

File: src/plugins/orangehrmAdminPlugin/Service/LocalizationService.php (hunk 4 above)
Comment (TharakaAthukorala) [reply src/plugins/orangehrmAdminPlugin/Service/LocalizationService.php:354]: need to discuss the max char length for source and target XLIFFs. If there is no limit we can remove this length validation.

File: src/plugins/orangehrmAdminPlugin/Service/LocalizationService.php [hunk 8]
```diff
@@ -230,23 +233,25 @@ private function getXliffXmlSources(array $i18nGroups, I18NLanguage $language):
         $root->setAttribute('srcLang', 'en_US');
         $root->setAttribute('trgLang', $language->getCode());
         $root->setAttribute('xmlns', 'urn:oasis:names:tc:xliff:document:2.0');
-        $root->setAttribute('date', @date('Y-m-d\TH:i:s\Z'));
 
         $file = $xml->createElement('file');
+        $file->setAttribute('id', 1);
```
Comment (RajithaKumara) [inline src/plugins/orangehrmAdminPlugin/Service/LocalizationService.php:238]: @devishke-orange  check whether we have better id for this

File: src/plugins/orangehrmAdminPlugin/Service/LocalizationService.php [hunk 9]
```diff
@@ -270,4 +275,165 @@ private function getXliffXmlSources(array $i18nGroups, I18NLanguage $language):
         }
         return $xml;
     }
+
+    public function symfonyXliffValidations(string $content, ?string $file = null): array
```
Comment (RajithaKumara) [inline src/plugins/orangehrmAdminPlugin/Service/LocalizationService.php:279]: add doc comment

File: src/plugins/orangehrmAdminPlugin/Service/LocalizationService.php [hunk 10]
```diff
@@ -270,4 +275,165 @@ private function getXliffXmlSources(array $i18nGroups, I18NLanguage $language):
...
+        foreach ($units as $unit) {
+            $unitId = $unit->getAttribute('id');
+            $source = $unit->getElementsByTagName('source')->item(0)->nodeValue;
+            $target = $unit->getElementsByTagName('target')->item(0)->nodeValue;
+
+            // Initialize an array to store errors for this unit
+            $unitErrors = "";
+
+            // Validate source text
+            if (strlen($source) > $maxLength) {
```
Comment (RajithaKumara) [inline src/plugins/orangehrmAdminPlugin/Service/LocalizationService.php:365]:
```suggestion
            if ($this->getTextHelper()->strLength($source, '8bit') > $maxLength) {
```
reference: https://github.com/orangehrm/orangehrm/blob/main/src/plugins/orangehrmCorePlugin/Api/V2/Validator/Rules/Base64Attachment.php#L98

File: src/plugins/orangehrmAdminPlugin/Service/LocalizationService.php [hunk 11]
```diff
@@ -270,4 +275,165 @@ private function getXliffXmlSources(array $i18nGroups, I18NLanguage $language):
...
+                $unitErrors .= "Source text exceeds maximum length of $maxLength characters.";
+            }
+
+            // Validate target text
+            if (strlen($target) > $maxLength) {
```
Comment (RajithaKumara) [inline src/plugins/orangehrmAdminPlugin/Service/LocalizationService.php:370]: `strlen`

**Review by RajithaKumara — CHANGES_REQUESTED** [review]
API tests are pending

**Review by devishke-orange — APPROVED** [review]

**Review by RajithaKumara — APPROVED** [review]

**Pull request merged**